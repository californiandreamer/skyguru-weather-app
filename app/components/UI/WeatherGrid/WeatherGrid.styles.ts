import { white } from 'app/constants/colors'
import { smallGap } from 'app/constants/values'
import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  weatherStatusRow: {
    paddingBottom: smallGap,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  weatherIcon: {
    paddingRight: smallGap,
    justifyContent: 'center',
    alignItems: 'center',
  },
  temperatureRow: {
    paddingVertical: smallGap,
    borderBottomColor: white,
    borderBottomWidth: 1,
    borderTopColor: white,
    borderTopWidth: 1,
  },
  temperatureValuesRow: {
    paddingTop: smallGap,
  },
  spinnerView: {
    maxHeight: 100,
  },
})
