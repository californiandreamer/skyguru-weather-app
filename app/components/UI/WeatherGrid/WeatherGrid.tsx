import StyledText from 'app/components/HOC/Text'
import Spinner from 'app/components/UI/Spinner/Spinner'
import {
  largeFontSize,
  pressOpacity,
  smallFontSize,
} from 'app/constants/values'
import { useTheme } from 'app/hooks'
import { showWeatherInfo } from 'app/store/actions/weatherInfo'
import { RootState } from 'app/store/reducers/rootReducer'
import { weatherFormatter } from 'app/utils/weatherFormatter'
import { weatherIconHandler } from 'app/utils/weatherIconHandler'
import React, { useEffect, useState } from 'react'
import { TouchableOpacity, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'

import styles from './WeatherGrid.styles'

export type WeatherT = {
  temperature?: number
  minTemperature?: number
  maxTemperature?: number
  weatherInfo?: string[]
  weatherStatus?: string[]
}

const WeatherGrid: React.FC<WeatherT> = (props) => {
  const dispatch = useDispatch()

  const [data, setData] = useState<WeatherT | null>(null)
  const [_, updateTheme] = useTheme()

  const { pending, currentWeather, error } = useSelector(
    (state: RootState) => state.currentWeather
  )

  const handleWeatherData = () => {
    const isPropsExist = Object.keys(props).length
    if (isPropsExist) {
      setData({ ...props })
    } else {
      if (currentWeather)
        setData({
          temperature: currentWeather?.main?.temp,
          minTemperature: currentWeather?.main?.temp_min,
          maxTemperature: currentWeather?.main?.temp_max,
          weatherInfo: currentWeather?.weather?.map((item) => item.main),
          weatherStatus: currentWeather?.weather?.map(
            (item) => item.description
          ),
        })
    }
  }

  useEffect(() => {
    handleWeatherData()
  }, [currentWeather, props])

  useEffect(() => {
    updateTheme({
      currentTime: currentWeather?.dt,
      sunrise: currentWeather?.sys.sunrise,
      sunset: currentWeather?.sys.sunset,
    })
  }, [currentWeather])

  const renderWeatherGrid = () => (
    <TouchableOpacity
      activeOpacity={pressOpacity}
      onPress={() =>
        currentWeather &&
        dispatch(
          showWeatherInfo({
            clouds: currentWeather.clouds.all,
            dew_point: 0,
            dt: currentWeather.dt,
            feels_like: { day: currentWeather.main.feels_like },
            humidity: currentWeather.main.humidity,
            pressure: currentWeather.main.pressure,
            sunrise: currentWeather.sys.sunrise,
            sunset: currentWeather.sys.sunset,
            temp: {
              day: currentWeather.main.temp,
              min: currentWeather.main.temp_min,
              max: currentWeather.main.temp_max,
            },
            uvi: currentWeather.id,
            visibility: currentWeather.visibility,
            weather: [...currentWeather.weather],
            wind_deg: currentWeather.wind.deg,
            wind_gust: currentWeather.wind.gust,
            wind_speed: currentWeather.wind.speed,
          })
        )
      }
    >
      <View style={styles.weatherStatusRow}>
        <View style={styles.weatherIcon}>
          {weatherIconHandler(
            data?.weatherStatus ? data?.weatherStatus[0] : undefined
          )}
        </View>
        <StyledText fontSize={smallFontSize}>{data?.weatherInfo}</StyledText>
      </View>
      <View style={styles.temperatureRow}>
        <StyledText fontSize={largeFontSize} lineHeight={80}>
          {weatherFormatter(data?.temperature, 'temperature')}
        </StyledText>
      </View>
      <View style={styles.temperatureValuesRow}>
        <StyledText fontSize={smallFontSize}>
          H: {weatherFormatter(data?.maxTemperature, 'temperature')} / L:{' '}
          {weatherFormatter(data?.minTemperature, 'temperature')}
        </StyledText>
      </View>
    </TouchableOpacity>
  )

  const renderSpinner = () => (
    <View style={styles.spinnerView}>
      <Spinner />
    </View>
  )

  return (
    <View style={styles.container}>
      {!pending && !error && data ? renderWeatherGrid() : renderSpinner()}
    </View>
  )
}

export default WeatherGrid
