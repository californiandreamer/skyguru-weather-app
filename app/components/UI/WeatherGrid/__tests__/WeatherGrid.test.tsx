import configureStore from 'app/store'
import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'

import WeatherGrid, { WeatherT } from '../WeatherGrid'

describe('WeatherGrid component', () => {
  it('renders correctly', () => {
    const store = configureStore()

    const weatherGridProps: WeatherT = {
      temperature: 16,
      minTemperature: 14,
      maxTemperature: 20,
      weatherInfo: ['Clear'],
      weatherStatus: ['clear sky'],
    }

    const tree = renderer
      .create(
        <Provider store={store}>
          <WeatherGrid {...weatherGridProps} />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
