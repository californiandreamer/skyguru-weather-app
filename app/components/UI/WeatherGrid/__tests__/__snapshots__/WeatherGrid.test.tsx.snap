// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherGrid component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "maxHeight": 100,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "transform": Array [
              Object {
                "rotateZ": "0deg",
              },
            ],
          }
        }
      >
        <SvgMock
          height={36}
          width={36}
        />
      </View>
    </View>
  </View>
</View>
`;
